atomics on CPU

I have the following simple atomic counter example that compiles and runs fine for CUDA, but doesn't compile on the CPU.
```
#include <iostream>
#include <CL/sycl.hpp>

namespace s = cl::sycl;

int main()
{
    s::queue q;

    int counter = 0;
    {
        s::buffer<int> counter_buf(&counter, 1);

        q.submit([&](cl::sycl::handler& cgh)
        {
            auto access_counter = counter_buf.get_access<cl::sycl::access::mode::read_write>(cgh);

            cgh.parallel_for<class atomic_increment>(s::range<1>(1 << 30), [=] (cl::sycl::id<1> tid)
            {
                s::atomic<int> atomic_counter { s::global_ptr<int> {&access_counter[0]} };
                atomic_counter.fetch_add(1);
            });
        });
    }

    std::cout << "Counter: " << counter << std::endl;
}
```

```
/opt/hipSYCL/CUDA/bin/syclcc-clang --hipsycl-gpu-arch=sm_52 -O3 atomic-counter.cpp -o atomic-counter-cuda
```
compiles and runs fine.

```
/opt/hipSYCL/CUDA/bin/syclcc-clang --hipsycl-platform=cpu -g atomic-counter.cpp -o atomic-counter-cpu
In file included from atomic-counter.cpp:3:
In file included from /opt/hipSYCL/CUDA/bin/../include/CL/sycl.hpp:58:
/opt/hipSYCL/CUDA/bin/../include/CL/sycl/atomic.hpp:103:12: error: use of undeclared identifier 'atomicAdd'
    return atomicAdd(_ptr, operand);
           ^
atomic-counter.cpp:23:32: note: in instantiation of function template specialization 'cl::sycl::atomic<int, cl::sycl::access::address_space::global_space>::fetch_add<int, nullptr>' requested here
                atomic_counter.fetch_add(1);
```

I know atomics are not fully supported (they are listed as a limitation in README), but other issues suggest that only minor features of atomics are missing. Is the above error by design, or is it something that can be easily fixed?

--- a/include/CL/sycl/atomic.hpp
+++ b/include/CL/sycl/atomic.hpp
@@ -68,11 +68,33 @@ public:
   /// \todo unimplemented
   HIPSYCL_KERNEL_TARGET
   void store(T operand, memory_order memoryOrder =
-      memory_order::relaxed) volatile;
+      memory_order::relaxed) volatile
+  {
+#ifdef __HIPSYCL_DEVICE_CALLABLE__
+#ifdef SYCL_DEVICE_ONLY
+    return detail::invalid_host_call_dummy_return<T>();     // really should be invalid_device_call
+#else
+    return __atomic_store_n(_ptr, operand, __ATOMIC_RELAXED);
+#endif
+#else
+    return detail::invalid_host_call_dummy_return<T>();
+#endif
+  }
 
   /// \todo unimplemented
   HIPSYCL_KERNEL_TARGET
-  T load(memory_order memoryOrder = memory_order::relaxed) const volatile;
+  T load(memory_order memoryOrder = memory_order::relaxed) const volatile
+  {
+#ifdef __HIPSYCL_DEVICE_CALLABLE__
+#ifdef SYCL_DEVICE_ONLY
+    return detail::invalid_host_call_dummy_return<T>();     // really should be invalid_device_call
+#else
+    return __atomic_load_n(_ptr, operand, __ATOMIC_RELAXED);
+#endif
+#else
+    return detail::invalid_host_call_dummy_return<T>();
+#endif
+  }
 
   HIPSYCL_KERNEL_TARGET
   T exchange(T operand, memory_order memoryOrder =

--- a/include/CL/sycl/atomic.hpp
+++ b/include/CL/sycl/atomic.hpp
@@ -86,7 +86,6 @@ public:
   }
 
   /* Available only when: T != float */
-  /// \todo unimplemented
   HIPSYCL_KERNEL_TARGET
   bool compare_exchange_strong(T &expected, T desired,
                                memory_order successMemoryOrder = memory_order::relaxed,
@@ -97,7 +96,7 @@ public:
     expected = atomicCAS(_ptr, expected, desired);
     return old == expected;
 #else
-    return detail::invalid_host_call_dummy_return<T>();
+    return __atomic_compare_exchange_n(_ptr, &expected, desired, false, __ATOMIC_RELAXED, __ATOMIC_RELAXED);
 #endif
   }
 

--- a/include/CL/sycl/atomic.hpp
+++ b/include/CL/sycl/atomic.hpp
@@ -90,7 +90,16 @@ public:
   HIPSYCL_KERNEL_TARGET
   bool compare_exchange_strong(T &expected, T desired,
                                memory_order successMemoryOrder = memory_order::relaxed,
-                               memory_order failMemoryOrder = memory_order::relaxed) volatile;
+                               memory_order failMemoryOrder = memory_order::relaxed) volatile
+  {
+#ifdef __HIPSYCL_DEVICE_CALLABLE__
+    T old = expected;
+    expected = atomicCAS(_ptr, expected, desired);
+    return old == expected;
+#else
+    return detail::invalid_host_call_dummy_return<T>();
+#endif
+  }
 
   /* Available only when: T != float */
   template<class t = T,

--- a/include/CL/sycl/atomic.hpp
+++ b/include/CL/sycl/atomic.hpp
@@ -65,36 +65,14 @@ public:
                   "Invalid pointer type for atomic<>");
   }
 
-  /// \todo unimplemented
+  /// \todo unimplemented on GPU
   HIPSYCL_KERNEL_TARGET
   void store(T operand, memory_order memoryOrder =
-      memory_order::relaxed) volatile
-  {
-#ifdef __HIPSYCL_DEVICE_CALLABLE__
-#ifdef SYCL_DEVICE_ONLY
-    return detail::invalid_host_call_dummy_return<T>();     // really should be invalid_device_call
-#else
-    return __atomic_store_n(_ptr, operand, __ATOMIC_RELAXED);
-#endif
-#else
-    return detail::invalid_host_call_dummy_return<T>();
-#endif
-  }
+      memory_order::relaxed) volatile;
 
-  /// \todo unimplemented
+  /// \todo unimplemented on GPU
   HIPSYCL_KERNEL_TARGET
-  T load(memory_order memoryOrder = memory_order::relaxed) const volatile
-  {
-#ifdef __HIPSYCL_DEVICE_CALLABLE__
-#ifdef SYCL_DEVICE_ONLY
-    return detail::invalid_host_call_dummy_return<T>();     // really should be invalid_device_call
-#else
-    return __atomic_load_n(_ptr, operand, __ATOMIC_RELAXED);
-#endif
-#else
-    return detail::invalid_host_call_dummy_return<T>();
-#endif
-  }
+  T load(memory_order memoryOrder = memory_order::relaxed) const volatile;
 
   HIPSYCL_KERNEL_TARGET
   T exchange(T operand, memory_order memoryOrder =
@@ -251,6 +229,23 @@ public:
 private:
   T* _ptr;
 };
+
+#ifdef HIPSYCL_PLATFORM_CPU
+  template <typename T, access::address_space addressSpace>
+  HIPSYCL_KERNEL_TARGET
+  void atomic<T,addressSpace>::store(T operand, memory_order) volatile
+  {
+    return __atomic_store_n(_ptr, operand, __ATOMIC_RELAXED);
+  }
+
+  template <typename T, access::address_space addressSpace>
+  HIPSYCL_KERNEL_TARGET
+  T atomic<T,addressSpace>::load(memory_order) const volatile
+  {
+    return __atomic_load_n(_ptr, __ATOMIC_RELAXED);
+  }
+#endif
+
 
 
 template <typename T, access::address_space addressSpace>

--- a/include/CL/sycl/atomic.hpp
+++ b/include/CL/sycl/atomic.hpp
@@ -82,6 +82,7 @@ public:
 #ifdef SYCL_DEVICE_ONLY
     return atomicExch(_ptr, operand);
 #else
+    assert(memoryOrder == memory_order::relaxed);
     return __atomic_exchange_n(_ptr, operand, __ATOMIC_RELAXED);
 #endif
 #else
@@ -101,6 +102,8 @@ public:
     expected = atomicCAS(_ptr, expected, desired);
     return old == expected;
 #else
+    assert(successMemoryOrder == memory_order::relaxed);
+    assert(failMemoryOrder == memory_order::relaxed);
     return __atomic_compare_exchange_n(_ptr, &expected, desired, false, __ATOMIC_RELAXED, __ATOMIC_RELAXED);
 #endif
 #else
@@ -119,6 +122,7 @@ public:
 #ifdef SYCL_DEVICE_ONLY
     return atomicAdd(_ptr, operand);
 #else
+    assert(memoryOrder == memory_order::relaxed);
     return __atomic_add_fetch(_ptr, operand, __ATOMIC_RELAXED);
 #endif
 #else
@@ -137,6 +141,7 @@ public:
 #ifdef SYCL_DEVICE_ONLY
     return atomicSub(_ptr, operand);
 #else
+    assert(memoryOrder == memory_order::relaxed);
     return __atomic_sub_fetch(_ptr, operand, __ATOMIC_RELAXED);
 #endif
 #else
@@ -155,6 +160,7 @@ public:
 #ifdef SYCL_DEVICE_ONLY
     return atomicAnd(_ptr, operand);
 #else
+    assert(memoryOrder == memory_order::relaxed);
     return __atomic_and_fetch(_ptr, operand, __ATOMIC_RELAXED);
 #endif
 #else
@@ -173,6 +179,7 @@ public:
 #ifdef SYCL_DEVICE_ONLY
     return atomicOr(_ptr, operand);
 #else
+    assert(memoryOrder == memory_order::relaxed);
     return __atomic_or_fetch(_ptr, operand, __ATOMIC_RELAXED);
 #endif
 #else
@@ -191,6 +198,7 @@ public:
 #ifdef SYCL_DEVICE_ONLY
     return atomicXor(_ptr, operand);
 #else
+    assert(memoryOrder == memory_order::relaxed);
     return __atomic_xor_fetch(_ptr, operand, __ATOMIC_RELAXED);
 #endif
 #else
@@ -251,15 +259,17 @@ private:
 #ifdef HIPSYCL_PLATFORM_CPU
   template <typename T, access::address_space addressSpace>
   HIPSYCL_KERNEL_TARGET
-  void atomic<T,addressSpace>::store(T operand, memory_order) volatile
+  void atomic<T,addressSpace>::store(T operand, memory_order memoryOrder) volatile
   {
+    assert(memoryOrder == memory_order::relaxed);
     return __atomic_store_n(_ptr, operand, __ATOMIC_RELAXED);
   }
 
   template <typename T, access::address_space addressSpace>
   HIPSYCL_KERNEL_TARGET
-  T atomic<T,addressSpace>::load(memory_order) const volatile
+  T atomic<T,addressSpace>::load(memory_order memoryOrder) const volatile
   {
+    assert(memoryOrder == memory_order::relaxed);
     return __atomic_load_n(_ptr, __ATOMIC_RELAXED);
   }
 #endif

--- a/include/CL/sycl/atomic.hpp
+++ b/include/CL/sycl/atomic.hpp
@@ -81,7 +81,7 @@ public:
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
     return atomicExch(_ptr, operand);
 #else
-    return detail::invalid_host_call_dummy_return<T>();
+    return __atomic_exchange_n(_ptr, operand, __ATOMIC_RELAXED);
 #endif
   }
 
@@ -102,7 +102,7 @@ public:
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
     return atomicAdd(_ptr, operand);
 #else
-    return detail::invalid_host_call_dummy_return<T>();
+    return __atomic_add_fetch(_ptr, operand, __ATOMIC_RELAXED);
 #endif
   }
 
@@ -116,7 +116,7 @@ public:
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
     return atomicSub(_ptr, operand);
 #else
-    return detail::invalid_host_call_dummy_return<T>();
+    return __atomic_sub_fetch(_ptr, operand, __ATOMIC_RELAXED);
 #endif
   }
 
@@ -130,7 +130,7 @@ public:
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
     return atomicAnd(_ptr, operand);
 #else
-    return detail::invalid_host_call_dummy_return<T>();
+    return __atomic_and_fetch(_ptr, operand, __ATOMIC_RELAXED);
 #endif
   }
 
@@ -144,7 +144,7 @@ public:
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
     return atomicOr(_ptr, operand);
 #else
-    return detail::invalid_host_call_dummy_return<T>();
+    return __atomic_or_fetch(_ptr, operand, __ATOMIC_RELAXED);
 #endif
   }
 
@@ -158,7 +158,7 @@ public:
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
     return atomicXor(_ptr, operand);
 #else
-    return detail::invalid_host_call_dummy_return<T>();
+    return __atomic_xor_fetch(_ptr, operand, __ATOMIC_RELAXED);
 #endif
   }
 

--- a/include/CL/sycl/atomic.hpp
+++ b/include/CL/sycl/atomic.hpp
@@ -224,7 +224,7 @@ public:
   template<class t = T,
            HIPSYCL_CONDITIONALLY_ENABLE_ATOMICS(t)>
   HIPSYCL_KERNEL_TARGET
-  T fetch_max(T operand, memory_order memoryOrder =
+  t fetch_max(t operand, memory_order memoryOrder =
       memory_order::relaxed) volatile
   {
 #ifdef __HIPSYCL_DEVICE_CALLABLE__

--- a/include/CL/sycl/atomic.hpp
+++ b/include/CL/sycl/atomic.hpp
@@ -79,9 +79,13 @@ public:
       memory_order::relaxed) volatile
   {
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
+#ifdef SYCL_DEVICE_ONLY
     return atomicExch(_ptr, operand);
 #else
     return __atomic_exchange_n(_ptr, operand, __ATOMIC_RELAXED);
+#endif
+#else
+    return detail::invalid_host_call_dummy_return<T>();
 #endif
   }
 
@@ -92,11 +96,15 @@ public:
                                memory_order failMemoryOrder = memory_order::relaxed) volatile
   {
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
+#ifdef SYCL_DEVICE_ONLY
     T old = expected;
     expected = atomicCAS(_ptr, expected, desired);
     return old == expected;
 #else
     return __atomic_compare_exchange_n(_ptr, &expected, desired, false, __ATOMIC_RELAXED, __ATOMIC_RELAXED);
+#endif
+#else
+    return detail::invalid_host_call_dummy_return<T>();
 #endif
   }
 
@@ -108,9 +116,13 @@ public:
       memory_order::relaxed) volatile
   {
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
+#ifdef SYCL_DEVICE_ONLY
     return atomicAdd(_ptr, operand);
 #else
     return __atomic_add_fetch(_ptr, operand, __ATOMIC_RELAXED);
+#endif
+#else
+    return detail::invalid_host_call_dummy_return<T>();
 #endif
   }
 
@@ -122,9 +134,13 @@ public:
       memory_order::relaxed) volatile
   {
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
+#ifdef SYCL_DEVICE_ONLY
     return atomicSub(_ptr, operand);
 #else
     return __atomic_sub_fetch(_ptr, operand, __ATOMIC_RELAXED);
+#endif
+#else
+    return detail::invalid_host_call_dummy_return<T>();
 #endif
   }
 
@@ -136,9 +152,13 @@ public:
       memory_order::relaxed) volatile
   {
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
+#ifdef SYCL_DEVICE_ONLY
     return atomicAnd(_ptr, operand);
 #else
     return __atomic_and_fetch(_ptr, operand, __ATOMIC_RELAXED);
+#endif
+#else
+    return detail::invalid_host_call_dummy_return<T>();
 #endif
   }
 
@@ -150,9 +170,13 @@ public:
       memory_order::relaxed) volatile
   {
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
+#ifdef SYCL_DEVICE_ONLY
     return atomicOr(_ptr, operand);
 #else
     return __atomic_or_fetch(_ptr, operand, __ATOMIC_RELAXED);
+#endif
+#else
+    return detail::invalid_host_call_dummy_return<T>();
 #endif
   }
 
@@ -164,9 +188,13 @@ public:
       memory_order::relaxed) volatile
   {
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
+#ifdef SYCL_DEVICE_ONLY
     return atomicXor(_ptr, operand);
 #else
     return __atomic_xor_fetch(_ptr, operand, __ATOMIC_RELAXED);
+#endif
+#else
+    return detail::invalid_host_call_dummy_return<T>();
 #endif
   }
 

--- a/include/CL/sycl/atomic.hpp
+++ b/include/CL/sycl/atomic.hpp
@@ -206,7 +206,15 @@ public:
       memory_order::relaxed) volatile
   {
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
+#ifdef SYCL_DEVICE_ONLY
     return atomicMin(_ptr, operand);
+#else
+    t old = load();
+    do
+    {
+        if (old < operand) return old;
+    } while(!compare_exchange_strong(old, operand));
+#endif
 #else
     return detail::invalid_host_call_dummy_return<T>();
 #endif
@@ -220,7 +228,15 @@ public:
       memory_order::relaxed) volatile
   {
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
+#ifdef SYCL_DEVICE_ONLY
     return atomicMax(_ptr, operand);
+#else
+    t old = load();
+    do
+    {
+        if (old > operand) return old;
+    } while(!compare_exchange_strong(old, operand));
+#endif
 #else
     return detail::invalid_host_call_dummy_return<T>();
 #endif

--- a/include/CL/sycl/atomic.hpp
+++ b/include/CL/sycl/atomic.hpp
@@ -214,6 +214,7 @@ public:
     {
         if (old < operand) return old;
     } while(!compare_exchange_strong(old, operand));
+    return operand;
 #endif
 #else
     return detail::invalid_host_call_dummy_return<T>();
@@ -236,6 +237,7 @@ public:
     {
         if (old > operand) return old;
     } while(!compare_exchange_strong(old, operand));
+    return operand;
 #endif
 #else
     return detail::invalid_host_call_dummy_return<T>();

